Tests for and refactoring of the depletion approximation

phoebe-p opened the pull request:
Added tests for all functions in analytic_solar_cells.depletion_approximation.py. Also added two fixtures to conftest.py which are used by these tests.

In addition, the code in depletion_approximation has been refactored to an extent (the functionality of qe_depletion and iv_depletion has not changed):

- get_j_top and get_j_bottom had identical functionality and have been replaced by a single function, get_j_dark
- depletion widths are calculated in a separate function
- the processing of the junction, where the order layer is checked and parameters are extracted, is the same for the QE and IV calculations and has thus been made into a separate function. This does have the downside that you are now getting a very long list of variables being returned from this function, @dalonsoa is this bad practice, or do you have a suggestion of a more elegant way of doing this?

The tests give 100% coverage of the module in its current state, including exceptions.

dalonsoa commented on tests/conftest.py:
```diff
@@ -15,3 +15,108 @@ def gauss_spectrum(wavelength):
+
+
+@fixture
+def np_junction():
```
This two fixtures are nearly identical and the only practical difference is the order of the dopants. I would put all of this (with some minor changes) in a normal function (not a `fixture`) just above this one, that takes as inputs the Na and Nd doping for the top and bottom layers:

```python
def junction(nd_top=0, na_top=0, nd_bottom=0, na_bottom=0)
```
Then the two fixtures just call this function with:

```python
@fixture
def np_junction()
    Nd=...
    Na=...
    return junction(nd_top=Nd, na_bottom=Na)

@fixture
def pn_junction()
    Nd=...
    Na=...
    return junction(nd_bottom=Nd, na_top=Na)
```

dalonsoa commented on tests/conftest.py:
```diff
@@ -15,3 +15,108 @@ def gauss_spectrum(wavelength):
+    prepare_solar_cell(test_junc, options)
+
+    solve_beer_lambert(test_junc, options)
+    return test_junc, light_source, options
```
`light_source` and `options` should be a `fixture` on their own that are passed as inputs to these ones.

phoebe-p commented:
Ok! After some experimentation, I figured out the tests were occasionally failing due to a very low Vbi being calculated at very low (< 50 K or so) temperatures. To be honest I'm not sure what the expected behaviour at those kinds of temperatures is but it is not really a use-case I can see being useful so I simply narrowed the range over which temperatures are generated. Just to make sure everything is ok I ran all the tests a few hundred times to see if they would ever fail and got 0 failures, obviously this is not a guarantee that they will never fail but even with all the random numbers they now seem very robust.

In addition:
- process_junction is split into two functions as suggested, one to identify the different regions and one to get the parameters out. Relevant tests also changed
- changed conftest.py: the pn_junction and np_junction fixtures are now both generated from a general 'junction' function and the light_source and options objects are generated separately.
